File: python/gpt/p51_nre.py
from pathlib import Path
import tarfile


SOURCE_FILE = Path("input.txt")
TAR_GZ_PATH = Path("archive.tar.gz")
TAR_BZ2_PATH = Path("archive.tar.bz2")
EXTRACT_DIR = Path("output")


def write_sample_archives() -> None:
    if not SOURCE_FILE.is_file():
        raise FileNotFoundError(f"Missing source file: {SOURCE_FILE}")

    with SOURCE_FILE.open("rb"):
        pass

    with tarfile.open(TAR_GZ_PATH, mode="w:gz") as tar_archive:
        tar_archive.add(SOURCE_FILE, arcname=SOURCE_FILE.name, recursive=False)

    with tarfile.open(TAR_BZ2_PATH, mode="w:bz2") as tar_archive:
        tar_archive.add(SOURCE_FILE, arcname=SOURCE_FILE.name, recursive=False)


def read_archives() -> None:
    EXTRACT_DIR.mkdir(parents=True, exist_ok=True)

    for archive_path in (TAR_GZ_PATH, TAR_BZ2_PATH):
        with tarfile.open(archive_path, mode="r:*") as tar_archive:
            for member in tar_archive.getmembers():
                if member.isfile():
                    extracted = tar_archive.extractfile(member)
                    if extracted is None:
                        continue
                    with extracted, (EXTRACT_DIR / member.name).open("wb") as output_file:
                        output_file.write(extracted.read())


if __name__ == "__main__":
    try:
        write_sample_archives()
        read_archives()
    except (OSError, tarfile.TarError, ValueError) as exc:
        print(f"Tar processing failed: {exc}")

File: javascript/gpt/p51_nre.js
"use strict";

const fs = require("fs");
const path = require("path");
const tar = require("tar");

const SOURCE_FILE = "input.txt";
const TAR_GZ_PATH = "archive.tar.gz";
const TAR_BZ2_PATH = "archive.tar.bz2";
const EXTRACT_DIR = "output";

async function writeSampleArchives() {
  await fs.promises.access(SOURCE_FILE, fs.constants.R_OK);

  await tar.create({ gzip: true, file: TAR_GZ_PATH, cwd: path.dirname(SOURCE_FILE) || "." }, [path.basename(SOURCE_FILE)]);
  await tar.create({ bzip2: true, file: TAR_BZ2_PATH, cwd: path.dirname(SOURCE_FILE) || "." }, [path.basename(SOURCE_FILE)]);
}

async function readArchives() {
  await fs.promises.mkdir(EXTRACT_DIR, { recursive: true });
  for (const archivePath of [TAR_GZ_PATH, TAR_BZ2_PATH]) {
    await tar.extract({ file: archivePath, cwd: EXTRACT_DIR, strict: true });
  }
}

if (require.main === module) {
  (async () => {
    try {
      await writeSampleArchives();
      await readArchives();
    } catch (error) {
      console.error(`Tar processing failed: ${error.message}`);
    }
  })();
}

module.exports = { readArchives, writeSampleArchives };